src/ray/object_manager/object_manager.cc

```diff
@@ -460,28 +460,58 @@ void ObjectManager::Push(const ObjectID &object_id, const ClientID &client_id) {
                    << ", total data size: " << data_size;
 
     UniqueID push_id = UniqueID::FromRandom();
-    for (uint64_t chunk_index = 0; chunk_index < num_chunks; ++chunk_index) {
-      rpc_service_.post([this, push_id, object_id, owner_address, client_id, data_size,
-                         metadata_size, chunk_index, rpc_client]() {
-        auto st = SendObjectChunk(push_id, object_id, owner_address, client_id, data_size,
-                                  metadata_size, chunk_index, rpc_client);
-        if (!st.ok()) {
-          RAY_LOG(WARNING) << "Send object " << object_id << " chunk failed due to "
-                           << st.message() << ", chunk index " << chunk_index;
-        }
-      });
-    }
+    SendChunkBatch(push_id, object_id, owner_address, client_id, data_size, metadata_size,
+                   0 /* from_chunk_index */, num_chunks, rpc_client);
   } else {
     // Push is best effort, so do nothing here.
     RAY_LOG(ERROR)
         << "Failed to establish connection for Push with remote object manager.";
   }
 }
 
+void ObjectManager::SendChunkBatch(const UniqueID &push_id, const ObjectID &object_id,
+                                   const rpc::Address &owner_address,
+                                   const ClientID &client_id, uint64_t data_size,
+                                   uint64_t metadata_size, uint64_t from_chunk_index,
+                                   uint64_t num_chunks,
+                                   std::shared_ptr<rpc::ObjectManagerClient> rpc_client) {
+  uint64_t max_batch = RayConfig::instance().object_manager_default_batch_size();
+  uint64_t n = 1;
+
+  uint64_t index = from_chunk_index;
+
+  while (true) {
```

Comment:
I found the logic in this loop a bit hard to follow. Could you add some comments?